CIPs/0017.md

```diff
@@ -0,0 +1,165 @@
+    `handleLogic`: If true, indicates the server should send back the best set of proofs/headers it has
+                   If false, the server should look at `requestedProofs` and send back proofs requested, with headers to fill gaps as appropriate.
+    `requestedProofs`: Set of proofs' metadata requested by the client
+  Server responds with `EpocHeadersOrProofs` dependent on `handleLogic`.
+  Both responses are subject to message size limits limiting the number of proofs/headers.
+EpochHeadersOrProofs: Contains the set of proofs & headers to help the client, as requested in `GetEpochHeadersOrProofs`.
+```
+
+Option 2: This option benefits from assumptions on the pattern of proof generation (likely at least one for every ~120 epochs) and by assuming proofs are well distributed (all full nodes have all proofs). The idea here is to pass a bitmap to the `GetEpochHeadersOrProofs` msg that indicates if the client knows the validator set for the past `SIZE_OF_BITMAP` (~240?) epochs. This allows the full node to optimize the set of proofs sent back, while minimizing computation for the client. The main win here would be in a case like this: the client uses a proof to sync to epoch 120, then syncs further to epoch 121, so it knows the validator set of 120 and 121. Then the client goes down for > 120 epochs, and when it comes back the server can figure out to send a proof from 120-240 that we assume is generated, rather than sending 119 epoch headers to verify.
+
+The msgs are:
+```
+GetEpochHeadersOrProofs:
+  Client sends (highestEpoch uint, knownValSets *big.Int)
+    `knownValSets` here would be a bitmap of the last ~240 or so epochs that the client knows the val set for
+  Server computes the optimal path of proofs & headers that the client can use to sync fastest and send them back using `EpochHeadersOrProofs`.
+EpochHeadersOrProofs:
+  Server sends this back from `GetEpochHeadersOrProofs` with the optimal set of proofs and headers this client can utilize.
+```  
+
+It may be preferred to combine both options, and send a bitmap with `handleLogic == false` in option 1.
+
+Regardless of the option chosen, the `les` client must then be modified to request proofs & headers during `lightest` sync. This will likely replace the function `getEpochOrNormalHeaders` within `eth/downloader`. This will also involve a new function to update the underlying validator set and blockchain data, which will be built to work with checkpoints as well.
+
+Upon receiving proofs and headers in a response message, the light client will verify headers the same way as before, and will verify proofs via a call into `celo-bls-go`'s `verify` fn.
+
+The verify function requires information on an epoch's validator set, so this will likely require some access changes (private -> public) and exposing more data structures to allow the validator set to be queried and modified upon proof verification.
+
+### Upgradeability
```

Comment:
One thing that might be worth covering is how we will handle gaps in our epoch -> validator set mapping. e.g. what happens if a lightest node wants to check if a transaction was included in a block from a previous epoch for which it doesn't have the validator set? IIRC you solved a similar problem at one rung lower in the validator set > block header > merkle proof hierarchy of the light client protocol.

 think it's fine to not implement the solution for now but it would be nice to know that the message design here will cover that case. (it seems like it does)

Reply:
Sorry, general public, to link a Slack thread you can't see, but some discussion on this in https://celo-org.slack.com/archives/CLPE043GT/p1597444936109200

Reply:
I think that's just about transaction inclusion, right? There are other reasons it's useful to be able to fetch and validate headers, so even if there were other ways to prove transaction inclusion the scheme I described above would still be useful

Reply:
Can you give me some examples of other reasons?

Reply:
Agreed the plan is for the protocol to be accommodating for this kind of request, but I won't prioritize the strict implementation of these requests yet as you mention